service/tracker/consensus.go

```diff
@@ -0,0 +1,303 @@
+// Copyright 2021 Optakt Labs OÜ
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package tracker
+
+import (
+	"fmt"
+
+	"github.com/dgraph-io/badger/v2"
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/storage/badger/operation"
+
+	"github.com/optakt/flow-dps/models/dps"
+)
+
+// Consensus is the DPS consensus follower, which uses a local protocol state
+// database to retrieve consensus-dependent data, while falling back on an
+// record holder to complement the rest of the data. It provides a callback for
+// the unstaked consensus follower on the Flow network that allows it to update
+// the cached data each time a block is finalized.
+// Consensus implements the `Chain` interface needed by the DPS indexer.
+type Consensus struct {
+	log  zerolog.Logger
+	db   *badger.DB
+	hold RecordHolder
+	last uint64
+}
+
+// NewConsensus returns a new instance of the DPS consensus follower, reading
+// from the provided protocol state database and the provided block record
+// holder.
+func NewConsensus(log zerolog.Logger, db *badger.DB, hold RecordHolder) (*Consensus, error) {
+
+	var root uint64
+	err := db.View(operation.RetrieveRootHeight(&root))
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve root height: %w", err)
+	}
```

Comment:
Are we always 100% guaranteed that when we create the consensus follower, the DB already contains the root height?